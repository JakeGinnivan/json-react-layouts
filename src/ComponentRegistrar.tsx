import React from 'react'
import { Logger } from 'typescript-log'

import { DataDefinition, ComponentState } from './DataLoading'
import { RouteBuilder } from './RouteBuilder'
import { Props } from './ComponentRenderer'

export interface ComponentRegistration<TType extends string, TProps extends {}, LoadDataServices> {
    type: TType
    render: RenderFunction<TProps, LoadDataServices>
    dataDefinition?: DataDefinition<any, any, LoadDataServices>
}

export function componentFactory<TLoadDataServices>() {
    return {
        /** Helper function to create the registration (to infer types) */
        createRegisterableComponent<TType extends string, TProps extends {}>(
            type: TType,
            render: RenderFunction<TProps, TLoadDataServices>,
        ): ComponentRegistration<TType, TProps, TLoadDataServices> {
            return { type, render }
        },

        /** Helper function to create the registration with data (to infer types) */
        createRegisterableComponentWithData<
            TType extends string,
            TProps extends {},
            TConfig extends {},
            TData
        >(
            type: TType,
            dataDefinition: DataDefinition<TConfig, TData, TLoadDataServices>,
            render: RenderFunction<
                TProps & ComponentState<TData> & { dataDefinitionArgs: TConfig },
                TLoadDataServices
            >,
        ): ComponentRegistration<
            TType,
            TProps & { dataDefinitionArgs: TConfig },
            TLoadDataServices
        > {
            // This is quite a complex transform which can't be modelled in typescript.
            //
            // The dataDefinition which is passed to this object is hidden from the types returned
            // The content area renderer has a data loader which will look for this property
            // Then use the loadData function
            const registrationWithData: any = { type, render, dataDefinition }
            // Once the data is loaded it will be passed to the render function on the
            // data prop, which will be typed as LoadedData<TData>

            // The route info looks like this:
            // { type: TType, props: TProps & { dataDefinition: TData } }
            return registrationWithData
        },
    }
}

/** A component definition inside route definitions */
export interface ComponentInformation<TType, TProps = {}> {
    type: TType
    props: TProps
}

export interface RenderFunctionServices<LoadDataServices> {
    routeBuilder: RouteBuilder<any, any, any, any>
    loadDataServices: LoadDataServices
}

/** The render function for components, converts the route props into a react component */
export type RenderFunction<TProps, LoadDataServices> = (
    props: TProps,
    services: RenderFunctionServices<LoadDataServices>,
) => React.ReactElement<any> | false | null

// Collection of errors that can be thrown by the ComponentRegistrar
export const Errors = {
    missing: (component: string) => `Component '${component}' has not been registered`,
}

/**
 * Allows regisration of components to render in composition content areas in
 * a type safe way
 *
 * @example ComponentRegistrar.register(myComponentRegistration)
 */
export class ComponentRegistrar<
    TLoadDataServices extends {},
    TComponents extends ComponentInformation<any> = never,
    TMiddlewareProps extends {} = {}
> {
    // the internal collection of registered components
    private registeredComponents: {
        [key: string]: ComponentRegistration<any, any, TLoadDataServices>
    } = {}
    private _componentMiddleware?: ComponentRendererMiddlewareRegistration<
        TLoadDataServices,
        TMiddlewareProps
    >
    public get componentMiddleware():
        | ComponentRendererMiddlewareRegistration<TLoadDataServices, TMiddlewareProps>
        | undefined {
        return this._componentMiddleware
    }

    constructor(public logger: Logger) {}

    isRegistered = (type: TComponents['type']) => {
        return this.registeredComponents[type] !== undefined
    }

    /**
     * Returns the data definition for the given component
     */
    getDataDefinition = (type: TComponents['type']) => {
        const foundComponent = this.registeredComponents[type]
        if (!foundComponent) {
            return undefined
        }
        return foundComponent.dataDefinition
    }

    /**
     * expects the type from T to be passed in as a parameter, from this we
     * can retrieve the render function associated with the component
     */
    get = (type: TComponents['type']) => {
        const foundComponent = this.registeredComponents[type]
        if (!foundComponent && process.env.NODE_ENV !== 'production') {
            // continue rendering in production only. otherwise throw, this is so the site does not crash
            // empty area will be rendered instead
            this.logger.warn(Errors.missing(type))
        }
        return foundComponent
    }

    /** used to register another component using ComponentRegistration<TType, TProps> */
    register<TType extends string, TProps extends {}>(
        registration: ComponentRegistration<TType, TProps, TLoadDataServices>,
    ): ComponentRegistrar<
        TLoadDataServices,
        Exclude<TComponents, never> | ComponentInformation<TType, TProps>
    > {
        if (this.registeredComponents[registration.type]) {
            throw new Error(`${registration.type} has already been registered`)
        }

        this.registeredComponents[registration.type] = registration

        return this as any
    }

    registerMiddleware<TRegistrationMiddlewareProps>(
        componentMiddleware: ComponentRendererMiddlewareRegistration<
            TLoadDataServices,
            TRegistrationMiddlewareProps
        >,
    ): ComponentRegistrar<
        TLoadDataServices,
        TComponents,
        TMiddlewareProps & TRegistrationMiddlewareProps
    > {
        // Figure out how to compose middlewares
        // if (this.componentMiddleware) {
        //     const previousMiddleware = this.componentMiddleware

        //     this.componentMiddleware = (props: any, next) => {
        //         return componentMiddleware(props, previousMiddleware)
        // }

        // This cast is safe because we are correctly typing the return type
        this._componentMiddleware = componentMiddleware as any

        return this as any
    }
}

export type ComponentRendererMiddlewareRegistration<TLoadDataServices, TMiddlewareProps> = (
    props: Props<TLoadDataServices> & TMiddlewareProps,
    services: RenderFunctionServices<TLoadDataServices>,
    next: RenderFunction<any, TLoadDataServices>,
) => React.ReactElement<any> | false | null
