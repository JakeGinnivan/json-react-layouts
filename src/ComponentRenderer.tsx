import React from 'react'
import { ComponentRegistrar, RenderFunctionServices, RenderFunction } from './ComponentRegistrar'
import { RouteBuilder } from './RouteBuilder'
import { ComponentState } from './DataLoading'

export type ComponentRendererMiddleware<TLoadDataServices> = (
    props: Props<TLoadDataServices>,
    services: RenderFunctionServices<TLoadDataServices>,
    next: RenderFunction<any, TLoadDataServices>,
) => React.ReactElement<any> | false | null

export interface Props<TLoadDataServices> {
    type: string
    routeBuilder: RouteBuilder<any, any, any, any>
    componentRegistrar: ComponentRegistrar<TLoadDataServices, any>
    componentProps: { componentRenderPath: string; dataDefinitionsArg?: any; [props: string]: any }
    loadDataServices: TLoadDataServices

    /** Allows a middleware to be specified for component rendering */
    renderComponentMiddleware?: ComponentRendererMiddleware<TLoadDataServices>
}

export const ComponentRenderer: React.FC<Props<any>> = props => {
    const component = props.componentRegistrar.get(props.type)
    if (component === undefined) {
        return null
    }

    const componentDataDefinition = props.componentRegistrar.getDataDefinition(props.type)

    const componentServices: RenderFunctionServices<any> = {
        loadDataServices: props.loadDataServices,
        routeBuilder: props.routeBuilder,
    }

    function render() {
        if (componentDataDefinition) {
            return (
                <props.routeBuilder.ComponentDataLoader
                    componentRegistrar={props.componentRegistrar}
                    componentRenderPath={props.componentProps.componentRenderPath}
                    dataDefinition={componentDataDefinition}
                    dataDefinitionArgs={props.componentProps.dataDefinitionArgs}
                    renderData={renderProps => {
                        if (!renderProps.lastAction.success) {
                            // We have failed to load data, use error boundaries
                            // to send error back up and render error page
                            throw renderProps.lastAction.error
                        }

                        const data: ComponentState<any> = renderProps.data.hasData
                            ? { data: { loaded: true, result: renderProps.data.result } }
                            : { data: { loaded: false } }

                        return (
                            component.render(
                                {
                                    ...props.componentProps,
                                    ...data,
                                },
                                componentServices,
                            ) || null
                        )
                    }}
                />
            )
        }
        return component.render(props.componentProps, componentServices) || null
    }

    if (props.renderComponentMiddleware) {
        return props.renderComponentMiddleware(props, componentServices, render) || null
    }
    return render()
}
ComponentRenderer.displayName = 'ComponentRenderer'
