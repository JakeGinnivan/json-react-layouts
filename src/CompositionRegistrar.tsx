import React from 'react'

import {
    ComponentInformation,
    Errors,
    ComponentRegistrar,
    componentFactory,
} from './ComponentRegistrar'
import { RouteBuilder } from './RouteBuilder'

import { ComponentRenderer } from './ComponentRenderer'

/** Creates a registerable composition, it is a two step process due to TypeScript limitations
 *
 * First createRegisterableComponent<'main'|'sidebar'>() creates a registration function
 * for a composition with two content areas, main and sidebar.
 * Second call that registration function to create the registration.
 */

export function createRegisterableComposition<TContentAreas extends string, TProps extends {}>() {
    return <TType extends string>(
        type: TType,
        render: CompositionRenderFunction<TContentAreas, TProps>,
    ): CompositionRegistration<TType, TContentAreas, TProps> => ({ type, render })
}

export interface CompositionRenderProps<TContentAreas, TProps, TLoadDataServices> {
    contentAreas: { [key in keyof TContentAreas]: React.ReactElement<any> }
    props: TProps

    /** Allows a middleware to be specified for component rendering */
    renderCompositionMiddleware?: (
        props: CompositionRenderProps<TContentAreas, TProps, TLoadDataServices>,
        next: RenderFunction<any, TLoadDataServices>,
    ) => React.ReactElement<any> | false | null
}

export interface CompositionInformation<
    TType extends string,
    TComponentInformation,
    TContentAreas extends string,
    TProps = {}
> extends ComponentInformation<TType, TProps> {
    type: TType
    props: TProps
    contentAreas: { [name in TContentAreas]: TComponentInformation[] }
}

export interface CompositionRegistration<TType, TContentAreas extends string, TProps = {}> {
    type: TType
    render: CompositionRenderFunction<TContentAreas, TProps>
}

export type RenderFunction<T extends ComponentInformation<any, TProps>, TProps = T['props']> = (
    props: TProps,
) => React.ReactElement<any>
export type CompositionRenderFunction<TContentAreas extends string, TProps> = (renderProps: {
    props: TProps
    contentAreas: { [key in TContentAreas]: React.ReactElement<any> }
}) => React.ReactElement<any>

export interface NestedCompositionProps {
    composition: CompositionInformation<any, any, any, any>
    componentRenderPath: string
}

export class CompositionRegistrar<
    TCompositions extends CompositionInformation<any, any, any>,
    TComponents extends ComponentInformation<any>,
    TLoadDataServices,
    TMiddlewareProps extends {}
> {
    /** Static constructor function due to type inference */
    static create<
        TRegisteredComponents extends ComponentInformation<any>,
        LoadDataServices,
        MiddlewareProps extends {}
    >(
        componentRegistrar: ComponentRegistrar<
            LoadDataServices,
            TRegisteredComponents,
            MiddlewareProps
        >,
    ) {
        const registrar = new CompositionRegistrar<
            never,
            | TRegisteredComponents
            | ComponentInformation<'nested-composition', NestedCompositionProps>,
            LoadDataServices,
            MiddlewareProps
        >(componentRegistrar as any)
        const { createRegisterableComponent } = componentFactory<LoadDataServices>()

        // Nested compositions, we register a composition renderer as a component
        registrar.componentRegistrar.register(
            createRegisterableComponent(
                'nested-composition',
                (props: NestedCompositionProps, services) => {
                    if (props.composition.type === 'nested-composition') {
                        throw new Error(
                            "nested-composition is registered as a component, it is not allowed within a nested composition as it's not a composition, please check the route",
                        )
                    }
                    return (
                        <registrar.CompositionRenderer
                            compositionInformation={props.composition}
                            componentRenderPath={`${props.componentRenderPath}/nested:${
                                props.composition.type
                            }`}
                            loadDataServices={services.loadDataServices}
                            // Think of this as a late bound type, right now it is invalid
                            // (no compositions registered), but the routebuilder cannot be constructed
                            // if this is the case
                            routeBuilder={services.routeBuilder as any}
                        />
                    )
                },
            ),
        )
        return registrar
    }

    private registeredCompositions: {
        [key: string]: {
            render: CompositionRenderFunction<any, any>
        }
    } = {}

    private constructor(
        public componentRegistrar: ComponentRegistrar<TLoadDataServices, TComponents>,
    ) {}

    get(type: TCompositions['type']) {
        const foundComponent = this.registeredCompositions[type]
        if (!foundComponent && process.env.NODE_ENV !== 'production') {
            // Warn a component is missing if not in production
            this.componentRegistrar.logger.warn(Errors.missing(type))
        }
        return foundComponent.render
    }

    registerComposition<TType extends string, TContentAreas extends string, TProps>(
        registration: CompositionRegistration<TType, TContentAreas, TProps>,
    ): CompositionRegistrar<
        | Exclude<TCompositions, never>
        | CompositionInformation<TType, TComponents, TContentAreas, TProps>,
        TComponents,
        TLoadDataServices,
        TMiddlewareProps
    > {
        if (this.registeredCompositions[registration.type]) {
            throw new Error(`${registration.type} has already been registered`)
        }

        this.registeredCompositions[registration.type] = {
            render: registration.render as any,
        }

        return this as any
    }

    CompositionRenderer: React.FC<
        CompositionRendererProps<
            TCompositions,
            TComponents & TMiddlewareProps,
            TLoadDataServices,
            TMiddlewareProps
        >
    > = (props): React.ReactElement<any> | null => {
        /**
         * The ContentAreaRenderer componentRenderPaths need to append `/[contentArea key]'
         * key as this logic is duped outside of react for the ssr
         */
        this.componentRegistrar.logger.debug(
            {
                componentRenderPath: props.componentRenderPath,
                type: props.compositionInformation.type,
            },
            'Rendering composition',
        )

        const contentAreas = Object.keys(props.compositionInformation.contentAreas).reduce<{
            [key: string]: React.ReactElement<any>
        }>((acc, val) => {
            acc[val] = (
                <this.ContentAreaRenderer
                    componentRenderPath={`${props.componentRenderPath}/${val}`}
                    contentArea={props.compositionInformation.contentAreas[val]}
                    routeBuilder={props.routeBuilder}
                    loadDataServices={props.loadDataServices}
                />
            )
            return acc
        }, {})
        const compositionProps: CompositionRenderProps<any, any, TLoadDataServices> = {
            contentAreas,
            props: props.compositionInformation.props,
        }
        const compositionElement = this.get(props.compositionInformation.type)
        return compositionElement(compositionProps)
    }

    ContentAreaRenderer: React.FC<
        ContentAreaRendererProps<TCompositions, TComponents & TMiddlewareProps, TLoadDataServices>
    > = (props): React.ReactElement<any> => {
        this.componentRegistrar.logger.debug(
            {
                componentRenderPath: props.componentRenderPath,
                components: props.contentArea.map(component => ({
                    type: component.type,
                })),
            },
            'Rendering content area',
        )

        return (
            <React.Fragment>
                {props.contentArea.map((item, index) => {
                    const { type, props: componentProps, ...middlewareProps } = item
                    return (
                        <ComponentRenderer
                            key={`${item.type}-${index}`}
                            type={type}
                            routeBuilder={props.routeBuilder}
                            componentRegistrar={this.componentRegistrar}
                            componentProps={{
                                ...componentProps,
                                componentRenderPath: `${props.componentRenderPath}[${index}]`,
                            }}
                            middlewareProps={middlewareProps}
                            loadDataServices={props.loadDataServices}
                            renderComponentMiddleware={this.componentRegistrar.componentMiddleware}
                        />
                    )
                })}
            </React.Fragment>
        )
    }
}

export interface CompositionRendererProps<
    TCompositions extends CompositionInformation<any, any, any>,
    TComponents extends ComponentInformation<any> & TMiddlewareProps,
    LoadDataServices,
    TMiddlewareProps extends {}
> {
    componentRenderPath: string
    compositionInformation: CompositionInformation<any, TComponents, any>
    routeBuilder: RouteBuilder<TCompositions, TComponents, LoadDataServices, TMiddlewareProps>
    loadDataServices: LoadDataServices
}

export interface ContentAreaRendererProps<
    TCompositions extends CompositionInformation<any, any, any>,
    TComponents extends ComponentInformation<any>,
    LoadDataServices
> {
    componentRenderPath: string
    contentArea: TComponents[]
    routeBuilder: RouteBuilder<TCompositions, TComponents, LoadDataServices, any>
    loadDataServices: LoadDataServices
}
