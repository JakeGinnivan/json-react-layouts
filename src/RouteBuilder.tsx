import { ComponentInformation } from './ComponentRegistrar'
import { CompositionInformation, CompositionRegistrar } from './CompositionRegistrar'
import { createPageRenderer, Props } from './PageRenderer'
import { ComponentState, LoadArguments } from './DataLoading'
import { DataLoaderResources, DataLoaderProps } from 'react-ssr-data-loader'
import { DataLoaderActions } from 'react-ssr-data-loader/dist/data-loader'

/**
 * This class is simply used to get better typescript completion and errors targeted to where they occur
 * its used to build up the route but also to render it
 */
export class RouteBuilder<
    TCompositions extends CompositionInformation<any, TComponents, any, any>,
    TComponents extends ComponentInformation<any> & TMiddlewareProps,
    LoadDataServices,
    TMiddlewareProps extends {}
> {
    _compositionType!: TCompositions
    _componentType!: TComponents & TMiddlewareProps
    _contentAreaType!: TComponents[]

    PageRenderer: React.FC<
        Props<TCompositions, TComponents, LoadDataServices, TMiddlewareProps>
    > = createPageRenderer(this)
    ComponentDataLoader: React.ReactType<
        DataLoaderProps<
            ComponentState<any>,
            DataLoaderActions<{}, ComponentState<any>>,
            LoadArguments<LoadDataServices>
        > &
            LoadArguments<LoadDataServices>
    >

    // Expects a composition registrar to be passed in
    constructor(
        public compositionRegistrar: CompositionRegistrar<
            TCompositions,
            TComponents,
            LoadDataServices,
            TMiddlewareProps
        >,
        resources: DataLoaderResources<LoadDataServices>,
    ) {
        this.ComponentDataLoader = resources.registerResource<any, LoadArguments<LoadDataServices>>(
            'component-data-loader',
            params => {
                return params.dataDefinition.loadData(params.dataDefinitionArgs, params)
            },
            ['componentRenderPath', 'dataDefinitionArgs'],
        )
    }

    component(component: TComponents & TMiddlewareProps) {
        return component
    }

    /** Nested compositions are special components
     * You need to use this function to create them then pass them as a component
     */
    nestedComposition(composition: TCompositions): ComponentInformation<any, any> {
        return {
            type: 'nested-composition',
            props: { composition },
        }
    }

    contentArea(...components: TComponents[]): Array<ComponentInformation<any>> {
        return components
    }

    composition(composition: TCompositions) {
        return composition
    }

    page(...compositions: TCompositions[]): TCompositions[] {
        return compositions
    }
}
