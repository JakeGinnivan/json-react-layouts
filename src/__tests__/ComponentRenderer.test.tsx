import React from 'react'
import { configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { ComponentRegistrar } from '../ComponentRegistrar'
import {
    testComponentRegistration,
    testComponentWithPropsRegistration,
    TestComponent,
    testCompositionRegistration,
    testComponentWithDataRegistration,
    TestComponentWithData,
} from '../testComponents'
import { CompositionRegistrar } from '../CompositionRegistrar'
import { RouteBuilder } from '../RouteBuilder'
import { ComponentRenderer } from '../ComponentRenderer'
import { consoleLogger } from 'typescript-log'
import { DataLoaderResources, DataProvider } from 'react-ssr-data-loader'

configure({ adapter: new Adapter() })
const logger = consoleLogger()

const registrar = new ComponentRegistrar(logger)
    .register(testComponentRegistration)
    .register(testComponentWithPropsRegistration)
const compositionRegistrar = CompositionRegistrar.create(registrar).registerComposition(
    testCompositionRegistration,
)
const routeBuilder = new RouteBuilder(compositionRegistrar, new DataLoaderResources())

it('can render simple test component', () => {
    const wrapper = mount(
        <ComponentRenderer
            type="test"
            componentProps={{ componentRenderPath: '/' }}
            routeBuilder={routeBuilder}
            componentRegistrar={registrar}
            loadDataServices={{}}
        />,
    )
    expect(wrapper.find(TestComponent).length).toBe(1)
})

it('can render simple component with renderProps', () => {
    const testTitle = 'testTheTitleHere'
    const renderer = (
        <ComponentRenderer
            type="testWithTitleProp"
            componentProps={{ title: testTitle, componentRenderPath: '/' }}
            routeBuilder={routeBuilder}
            componentRegistrar={registrar}
            loadDataServices={{}}
        />
    )
    const wrapper = mount(renderer)
    expect(wrapper.text()).toContain(testTitle)
})

it('can load data for component', async () => {
    const componentRegistrar = new ComponentRegistrar(logger).register(
        testComponentWithDataRegistration,
    )
    const compositionRegisrar = CompositionRegistrar.create(componentRegistrar).registerComposition(
        testCompositionRegistration,
    )
    const resources = new DataLoaderResources<{}>()
    const routeBuilder = new RouteBuilder(compositionRegisrar, resources)

    const wrapper = mount(
        <DataProvider resources={resources} globalProps={{}}>
            <compositionRegisrar.ContentAreaRenderer
                componentRenderPath="test"
                contentArea={[
                    { type: 'test-with-data', props: { dataDefinition: { dataArg: 'Foo' } } },
                ]}
                routeBuilder={routeBuilder}
                loadDataServices={{}}
            />
        </DataProvider>,
    )

    expect(wrapper.find(TestComponentWithData).text()).toBe('Loading')
    await new Promise(resolve => setTimeout(resolve))
    await new Promise(resolve => setTimeout(resolve))

    const component = wrapper.update().find(TestComponentWithData)
    expect(component.text()).toBe('Length: 3')
    expect(component.props()).toMatchSnapshot()
})
