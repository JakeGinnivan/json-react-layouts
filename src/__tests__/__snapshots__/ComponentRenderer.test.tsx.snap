// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can load data for component 1`] = `
Object {
  "componentRenderPath": "test[0]",
  "data": Object {
    "loaded": true,
    "result": 3,
  },
  "dataDefinitionArgs": Object {
    "dataArg": "Foo",
  },
  "length": 3,
}
`;
